efi/image_rules.go

```diff
@@ -0,0 +1,243 @@
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package efi
+
+import (
+	"bytes"
+	"crypto"
+	"fmt"
+
+	"golang.org/x/xerrors"
+)
+
+type (
+	newImageLoadHandlerFn func(peImageHandle) (imageLoadHandler, error)
+)
+
+// imagePredicate is used for testing image properties.
+type imagePredicate interface {
+	Matches(image peImageHandle) (bool, error)
+}
+
+type imagePredicateAny []imagePredicate
+
+// imageMatchesAny returns a predicate that is satisfied if any of
+// the supplied predicates are satisfied.
+func imageMatchesAny(predicates ...imagePredicate) imagePredicate {
+	return imagePredicateAny(predicates)
+}
+
+func (p imagePredicateAny) Matches(image peImageHandle) (bool, error) {
+	for _, pred := range p {
+		matches, err := pred.Matches(image)
+		if err != nil {
+			return false, err
+		}
+		if matches {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+type imagePredicateAll []imagePredicate
+
+// imageMatchesAll returns a predicate that is satisfied if all of
+// the supplied predicates are satisfied.
+func imageMatchesAll(predicates ...imagePredicate) imagePredicate {
+	return imagePredicateAll(predicates)
+}
+
+func (p imagePredicateAll) Matches(image peImageHandle) (bool, error) {
+	for _, pred := range p {
+		matches, err := pred.Matches(image)
+		if err != nil {
+			return false, err
+		}
+		if !matches {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// imageRule is a single rule associated with a set of rules.
+type imageRule struct {
+	name   string
+	match  imagePredicate
+	create newImageLoadHandlerFn
+}
+
+// newImageRule returns a new imageRule with the supplied arguments.
+func newImageRule(name string, match imagePredicate, create newImageLoadHandlerFn) *imageRule {
+	return &imageRule{
+		name:   name,
+		match:  match,
+		create: create,
+	}
+}
+
+// imageRules is used to construct an imageLoadHandler from a peImageHandle
+// using a set of rules.
+type imageRules struct {
+	name  string
+	rules []*imageRule
+}
+
+func (r *imageRules) String() string {
+	return r.name + " image rules"
+}
+
+// NewImageLoadHandler implements imageLoadHandlerConstructor
+func (r *imageRules) NewImageLoadHandler(image peImageHandle) (imageLoadHandler, error) {
+	for _, rule := range r.rules {
+		matches, err := rule.match.Matches(image)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot run \"%s\" rule: %w", rule.name, err)
```

Comment:
should s/rule/image rule/ ? or will it be clear somehow from context?

Reply:
I've updated this.